src/main/java/com/netdocuments/connect/kafka/handler/source/NDSourceHandler.java

```diff
@@ -109,99 +163,292 @@ public SourceRecordBuilder handle(SourceHandlerParams params) {
-          return false;
-        }
-      }
-      newValue.put("event", type.schemaName());
-      newValue.put("key", docEvent.key());
     } else {
       LOGGER.warn("unexpected event type {}", type);
       return false;
     }
 
     try {
-      byte[] value = objectMapper.writeValueAsBytes(newValue);
-      builder.value(null, value);
+      builder.value(null, convertToBytes(newValue, docEvent));
       return true;
-    } catch (JsonProcessingException e) {
+    } catch (DataException e) {
+      LOGGER.error("Failed to serialize data", e);
+      return false;
+    }
+  }
+
+  /**
+   * Creates a value map for deletion or expiration events.
+   */
+  private Map<String, Object> createDeletionOrExpirationValue(DocumentEvent docEvent, DocumentEvent.Type type) {
+    Map<String, Object> newValue = new HashMap<>();
+    newValue.put("event", type.schemaName());
+    newValue.put("key", docEvent.key());
+    return newValue;
+  }
+
+  /**
+   * Creates a value map for mutation events, extracting specified fields.
+   */
+  private Map<String, Object> createMutationValue(DocumentEvent docEvent, byte[] content) {
+    try {
+      Map<String, Object> newValue = JsonPropertyExtractor.extract(new ByteArrayInputStream(content),
+          fields.toArray(new String[fields.size()]));
+      newValue.put("event", DocumentEvent.Type.MUTATION.schemaName());
+      newValue.put("key", docEvent.key());
+      return newValue;
+    } catch (Exception e) {
+      LOGGER.error("Error while extracting fields from document", e);
+      return null;
+    }
+  }
+
+  /**
+   * Converts a value to bytes, applying CloudEvent format if necessary.
+   */
+  private byte[] convertToBytes(Map<String, Object> value, DocumentEvent docEvent) {
+    if (!cloudevent) {
+      return serializeToJson(value);
+    }
+    return withCloudEvent(serializeToJson(value), docEvent);
+  }
+
+  /**
+   * Serializes an object to JSON bytes.
+   */
+  private byte[] serializeToJson(Object value) {
+    try {
+      return OBJECT_MAPPER.writeValueAsBytes(value);
+    } catch (IOException e) {
       throw new DataException("Failed to serialize data", e);
     }
   }
+
+  /**
+   * Wraps the given value in a CloudEvent format.
+   */
+  private byte[] withCloudEvent(byte[] value, DocumentEvent documentEvent) {
+    Map<String, Object> cloudEventData = createCloudEventData(documentEvent);
+    byte[] cloudEventBytes = serializeToJson(cloudEventData);
+
+    ByteArrayBuilder result = new ByteArrayBuilder(
+        cloudEventBytes.length + ",\"data\":".getBytes().length + value.length)
+        .append(cloudEventBytes, cloudEventBytes.length - 1)
+        .append(",\"data\":".getBytes())
+        .append(value)
+        .append((byte) '}');
+    return result.build();
+  }
+
+  /**
+   * Creates the CloudEvent metadata for a document event.
+   */
+  private Map<String, Object> createCloudEventData(DocumentEvent documentEvent) {
```

Comment:
Is there a cloudevent library you could use for this?

Reply:
It makes the code significantly more complicated as well as adding an additional library.